### ISSUE TITLE
Does fairseq-train shuffle data within a batch?

### ISSUE THREAD

**Issue by ylmeng** [pytorch/fairseq#2250]
I understand batches are shuffled, unless "--curriculum" is used in some conditions.
However, within a batch, are the data points shuffled as well? If so, is it possible to fix it?

**Comment by alexeib** [comment]
it is not currently shuffled if i remember correctly. PRs welcome with an option to shuffle

**Comment by myleott** [comment]
To clarify, the order of batches are shuffled between epochs, but the items within a given batch are the same across epochs.

If you wanted to reconstruct the batches at each epoch, you could potentially look at ResamplingDataset: https://github.com/pytorch/fairseq/blob/master/fairseq/data/resampling_dataset.py.

Note that for tasks like Translation, to fill up `--max-tokens` most efficiently, we sort examples by length before batching. So you'd need to consider how you want to introduce randomness.

**Comment by ylmeng** [comment]
Thanks. In my experiment I did some data augmentation in which each input-output data point is altered and added to the training set. I hoped the altered data point can immediately follow the original data point, because that will make my life easier for some experiments. However, I figured the order within a batch is not the same as in training data file.
It's still good to know that the content in a specific batch will not change across epochs.

**Comment by ereday** [comment]
@myleott Is sorting handled inside fairseq-preprocess or fairseq-train? Also, is there a way to change the way batches created? For example, disabling sorting mechanism and/or using a custom function to decide which examples goes into the same batch?

**Comment by ereday** [comment]
@ylmeng Did you manage to make altered data point follow the original point in the batch? If so, could you please let me know how did you do that? I'm trying to do something similar as well.
